docs/design/2024-09-06-maintain-priority-queue-in-memory.md

```diff
@@ -0,0 +1,556 @@
+  - [Current Implementation](#current-implementation)
+    - [Calculate the score](#calculate-the-score)
+  - [Detailed Design](#detailed-design)
+    - [Goals](#goals)
+    - [Update the priority queue based on DML and DDL changes](#update-the-priority-queue-based-on-dml-and-ddl-changes)
+    - [Incremental Update](#incremental-update)
+      - [DML](#dml)
+      - [DDL](#ddl)
+    - [The New Priority Queue - Maintain the heap in memory](#the-new-priority-queue---maintain-the-heap-in-memory)
+      - [Information to be stored](#information-to-be-stored)
+      - [Basic API](#basic-api)
+      - [Reflect on DML Changes](#reflect-on-dml-changes)
+      - [Reflect on DDL Changes](#reflect-on-ddl-changes)
+        - [Golang-type view from the subscriber](#golang-type-view-from-the-subscriber)
+        - [Golang-type view from the publisher](#golang-type-view-from-the-publisher)
+        - [Persistence view from publisher](#persistence-view-from-publisher)
+        - [Subscribe API behaviors](#subscribe-api-behaviors)
+        - [Implementation of subscribing](#implementation-of-subscribing)
+        - [Where does it run?](#where-does-it-run)
+        - [Future possibility](#future-possibility)
+      - [Reflect on Time Change](#reflect-on-time-change)
+      - [Failover](#failover)
+  - [Test Design](#test-design)
+    - [Functional Tests](#functional-tests)
+    - [Scenario Tests](#scenario-tests)
+    - [Compatibility Tests](#compatibility-tests)
+    - [Benchmark Tests](#benchmark-tests)
+  - [Impacts \& Risks](#impacts--risks)
+  - [Investigation \& Alternatives](#investigation--alternatives)
+  - [Unresolved Questions](#unresolved-questions)
+
+## Introduction
+
+This design proposes a new priority queue implementation that maintains a heap in memory to improve performance and reduce the overhead of rebuilding the queue.
+
+## Motivation or Background
+
+Currently, we use a priority queue at TiDB to automatically update statistics. However, we currently maintain this queue by rebuilding it every time. This causes the following problems:
+
+1. After the introduction of information schema version 2, TiDB no longer stores all table information in memory. As a result, the priority queue must retrieve table information from TiKV, which can be slow when dealing with a large number of tables. Therefore, it's important to prevent any operations in the priority queue that require loading all table information.
+
+2. Because we scan it too frequently, we'll spend a lot of CPU on building this queue, since we'll be scanning all the stats cache and the schema cache. See more: <https://github.com/pingcap/tidb/issues/49972>
+We can try to keep the priority queue in memory to solve the above issues instead of rebuilding it every time.
+
+## Current Implementation
+
+### Calculate the score
+
+In the current implementation, we use a weighted scoring approach to maintain the queue, which is our current scoring criteria.
+
+Weight table:
```

Comment:
I'm not sure we should call it "weight" or "scroe". Weight means it will be multiplied to another number. But seems we directly add them up?

Reply:
The final formula is:
```go
//	priority_score = (0.6 * math.Log10(1 + ChangeRatio) +
//	                  0.1 * (1 - math.Log10(1 + TableSize)) +
//	                  0.3 * math.Log10(1 + math.Sqrt(AnalysisInterval)) +
//	                  special_event[event])
```

Reply:
Updated.